Non-waiting algorithm #53

ongrid opened the pull request:
Calculate the epoch that is both
finalized on ETH2 and reportable on ETH1

vshvsh commented on app/oracle.py:
```diff
@@ -167,50 +166,33 @@ def prompt(prompt_message, prompt_end):
+    logging.info(f'Potentially reportable epoch: {potentially_reportable_epoch} (from ETH1 contract)')
+    finalized_epoch_beacon = beacon.get_finalized_epoch()
+    logging.info(f'Last finalized epoch: {finalized_epoch_beacon} (from Beacon)')
+    reportable_epoch = min(finalized_epoch_beacon, potentially_reportable_epoch) // epochs_per_frame * epochs_per_frame
```
I'd frame it as 
reportable_epoch = `min(potentially_reportable_epoch, (finalized_epoch_beacon // epochs_per_frame) * epochs_per_frame)` which amounts to the same thing but makes sure SC is the main source of truth
After that, it needs a check that it's strictly greater than the already reported epoch, so that we won't run all the code down there when it'll cleary fail on submission.

vshvsh commented on app/oracle.py:
```diff
@@ -167,50 +166,33 @@ def prompt(prompt_message, prompt_end):
-    logging.info(f'Lido validators on Beacon chain: {validators_on_beacon}')
+
+    logging.info(f'Total balance on Beacon: {sum_balance} wei')
+    logging.info(f'Lido validators on Beacon: {validators_on_beacon}')
```
these numbers alone are not enough to understand if the report looks normal-ish or not for an operator
let's add the time between the last reported epoch and currently reported, and profit/shashing calculations there, at a minimum

ongrid commented:
Since the implementation of in-contract getter #231 of last pushed frame got postponed, the first mainnet release of the daemon v0.2x will rely on logs fetching

```
In:
oracle.events.Completed.getLogs(fromBlock=0, toBlock='latest')[-1]

Out:
AttributeDict({'args': AttributeDict({'epochId': 100,
  'beaconBalance': 576959499674000000000,
  'beaconValidators': 18}),
 'event': 'Completed',
 'logIndex': 0,
 'transactionIndex': 0,
 'transactionHash': HexBytes('0xaedf2a4bb4dc349bd07ac7a3cd4ecd7159d5eec799cb4b2d4b499b1808a16d36'),
 'address': '0xcD3db5ca818a645359e09543Cc0e5b7bB9593229',
 'blockHash': HexBytes('0x497196e6c86b75d416e311e0859f69de8100214ec3c9cf1e1e5a84fa8c6cd478'),
 'blockNumber': 1060})
```

ongrid commented:
Current output. @vshvsh could you check please?

```
    INFO 2020-12-09 19:27:00 <daemon> Connected to E2E (PoA engine)
    INFO 2020-12-09 19:27:00 <daemon> Injecting PoA compatibility middleware
    INFO 2020-12-09 19:27:00 <daemon> MEMBER_PRIV_KEY provided, running in transactable (PRODUCTION) mode
    INFO 2020-12-09 19:27:00 <daemon> Member account: 0xF2De2b2dD8881DE68390Da81EDc6Ab93c8e13291
    INFO 2020-12-09 19:27:00 <daemon> DAEMON=0 Running in single iteration mode (will exit after reporting).
    INFO 2020-12-09 19:27:00 <daemon> ETH1_NODE=http://127.0.0.1:8545
    INFO 2020-12-09 19:27:00 <daemon> BEACON_NODE=http://127.0.0.1:5052 (Lighthouse API)
    INFO 2020-12-09 19:27:00 <daemon> SLEEP=60 s
    INFO 2020-12-09 19:27:00 <daemon> GAS_LIMIT=1500000 gas units
    INFO 2020-12-09 19:27:00 <daemon> POOL_CONTRACT=0xc12e8e7adcEaF31c1Ca5F8aFD99AB88439628183
    INFO 2020-12-09 19:27:00 <daemon> Oracle contract address: 0xcD3db5ca818a645359e09543Cc0e5b7bB9593229 (auto-discovered)
    INFO 2020-12-09 19:27:00 <daemon> Registry contract address: 0x7fAF80E96530e5cd13a1f35701fcc6b334B2FD75 (auto-discovered)
    INFO 2020-12-09 19:27:00 <daemon> Seconds per slot: 1 (auto-discovered)
    INFO 2020-12-09 19:27:00 <daemon> Slots per epoch: 8 (auto-discovered)
    INFO 2020-12-09 19:27:00 <daemon> Epochs per frame: 20 (auto-discovered)
    INFO 2020-12-09 19:27:00 <daemon> Genesis time: 1607529992 (auto-discovered)
    INFO 2020-12-09 19:27:00 <daemon> Starting the main loop
    INFO 2020-12-09 19:27:00 <daemon> Getting last pushed frame
    INFO 2020-12-09 19:27:00 <daemon> Previously reported epoch: 120 (from ETH1 event)
    INFO 2020-12-09 19:27:00 <daemon> Previously reported beaconBalance: 96230134709000000000 wei or 96.230134709 ETH (from ETH1 event)
    INFO 2020-12-09 19:27:00 <daemon> Previously reported beaconValidators: 3 (from ETH1 event)
    INFO 2020-12-09 19:27:00 <daemon> Timestamp of previous report: 2020-12-09 19:23:21 or 1607531001 (from ETH1 event)
    INFO 2020-12-09 19:27:00 <daemon> Potentially reportable epoch: 140 (from ETH1 contract)
    INFO 2020-12-09 19:27:00 <daemon> Last finalized epoch: 151 (from Beacon)
    INFO 2020-12-09 19:27:00 <daemon> Reportable state: epoch:140 slot:1120
    INFO 2020-12-09 19:27:00 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-09 19:27:00 <daemon> Node operators in registry: 2
    INFO 2020-12-09 19:27:00 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-09 19:27:00 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-09 19:27:00 <daemon> Total validator keys in registry: 20
    INFO 2020-12-09 19:27:00 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-09 19:27:01 <daemon> Validator balances on beacon for slot: 1120
    INFO 2020-12-09 19:27:01 <daemon> Pubkey: 0x8e7ebb0d21 Balance: 32098355460 Gwei
    INFO 2020-12-09 19:27:01 <daemon> Pubkey: 0xa3d9511870 Balance: 32097926961 Gwei
    INFO 2020-12-09 19:27:01 <daemon> Pubkey: 0x94f9f5c1ad Balance: 32097514542 Gwei
    INFO 2020-12-09 19:27:01 <daemon> Validator balances sum: 96293796963 Gwei
    INFO 2020-12-09 19:27:01 <daemon> Total balance on Beacon: 96293796963000000000 wei
    INFO 2020-12-09 19:27:01 <daemon> Validators on Beacon: 3
    INFO 2020-12-09 19:27:01 <daemon> Time since previous report: 0:03:40 or 220 s
    INFO 2020-12-09 19:27:01 <daemon> Balance change since previus report: +63662254000000000 wei or +0.063662254 ETH
    INFO 2020-12-09 19:27:01 <daemon> Validators change since previus report: +0
    INFO 2020-12-09 19:27:01 <daemon> Tx call data: oracle.reportBeacon(140, 96293796963000000000, 3)
    INFO 2020-12-09 19:27:01 <daemon> Calling tx locally is succeeded.
Tx data: {'value': 0, 'gasPrice': 1000000000, 'chainId': 1337, 'from': '0xF2De2b2dD8881DE68390Da81EDc6Ab93c8e13291', 'gas': 1500000, 'to': '0xcD3db5ca818a645359e09543Cc0e5b7bB9593229', 'data': '0x62eeb732000000000000000000000000000000000000000000000000000000000000008c00000000000000000000000000000000000000000000000538584a26c5341e000000000000000000000000000000000000000000000000000000000000000003'}
Should we sent this TX? [y/n]: wtf
Please respond with [y or n]: 
Please respond with [y or n]: y
    INFO 2020-12-09 19:27:30 <daemon> Preparing TX... CTRL-C to abort
    INFO 2020-12-09 19:27:33 <daemon> TX hash: 0x18dd796036542a0c25c173ac0b6a8b628e5834277db4ee94cfae0f0c807c2d73 ... CTRL-C to abort
    INFO 2020-12-09 19:27:36 <daemon> Sending TX... CTRL-C to abort
    INFO 2020-12-09 19:27:39 <daemon> TX has been sent. Waiting for receipt...
    INFO 2020-12-09 19:27:45 <daemon> TX successful
    INFO 2020-12-09 19:27:45 <daemon> We are in single-iteration mode, so exiting. Set DAEMON=1 env to run in the loop.
```

ongrid commented:
With negative response

```
    INFO 2020-12-09 19:35:06 <daemon> Connected to E2E (PoA engine)
    INFO 2020-12-09 19:35:06 <daemon> Injecting PoA compatibility middleware
    INFO 2020-12-09 19:35:06 <daemon> MEMBER_PRIV_KEY provided, running in transactable (PRODUCTION) mode
    INFO 2020-12-09 19:35:06 <daemon> Member account: 0xF2De2b2dD8881DE68390Da81EDc6Ab93c8e13291
    INFO 2020-12-09 19:35:06 <daemon> DAEMON=0 Running in single iteration mode (will exit after reporting).
    INFO 2020-12-09 19:35:06 <daemon> ETH1_NODE=http://127.0.0.1:8545
    INFO 2020-12-09 19:35:06 <daemon> BEACON_NODE=http://127.0.0.1:5052 (Lighthouse API)
    INFO 2020-12-09 19:35:06 <daemon> SLEEP=60 s
    INFO 2020-12-09 19:35:06 <daemon> GAS_LIMIT=1500000 gas units
    INFO 2020-12-09 19:35:06 <daemon> POOL_CONTRACT=0xc12e8e7adcEaF31c1Ca5F8aFD99AB88439628183
    INFO 2020-12-09 19:35:06 <daemon> Oracle contract address: 0xcD3db5ca818a645359e09543Cc0e5b7bB9593229 (auto-discovered)
    INFO 2020-12-09 19:35:06 <daemon> Registry contract address: 0x7fAF80E96530e5cd13a1f35701fcc6b334B2FD75 (auto-discovered)
    INFO 2020-12-09 19:35:06 <daemon> Seconds per slot: 1 (auto-discovered)
    INFO 2020-12-09 19:35:06 <daemon> Slots per epoch: 8 (auto-discovered)
    INFO 2020-12-09 19:35:06 <daemon> Epochs per frame: 20 (auto-discovered)
    INFO 2020-12-09 19:35:06 <daemon> Genesis time: 1607529992 (auto-discovered)
    INFO 2020-12-09 19:35:06 <daemon> Starting the main loop
    INFO 2020-12-09 19:35:06 <daemon> Getting last pushed frame
    INFO 2020-12-09 19:35:06 <daemon> Previously reported epoch: 140 (from ETH1 event)
    INFO 2020-12-09 19:35:06 <daemon> Previously reported beaconBalance: 96293796963000000000 wei or 96.29379696299999 ETH (from ETH1 event)
    INFO 2020-12-09 19:35:06 <daemon> Previously reported beaconValidators: 3 (from ETH1 event)
    INFO 2020-12-09 19:35:06 <daemon> Timestamp of previous report: 2020-12-09 19:27:45 or 1607531265 (from ETH1 event)
    INFO 2020-12-09 19:35:06 <daemon> Potentially reportable epoch: 200 (from ETH1 contract)
    INFO 2020-12-09 19:35:06 <daemon> Last finalized epoch: 212 (from Beacon)
    INFO 2020-12-09 19:35:06 <daemon> Reportable state: epoch:200 slot:1600
    INFO 2020-12-09 19:35:06 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-09 19:35:06 <daemon> Node operators in registry: 2
    INFO 2020-12-09 19:35:06 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-09 19:35:07 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-09 19:35:07 <daemon> Total validator keys in registry: 20
    INFO 2020-12-09 19:35:07 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-09 19:35:07 <daemon> Validator balances on beacon for slot: 1600
    INFO 2020-12-09 19:35:07 <daemon> Pubkey: 0xa3d9511870 Balance: 32150105406 Gwei
    INFO 2020-12-09 19:35:07 <daemon> Pubkey: 0x8e7ebb0d21 Balance: 32149252497 Gwei
    INFO 2020-12-09 19:35:07 <daemon> Pubkey: 0x94f9f5c1ad Balance: 32153158264 Gwei
    INFO 2020-12-09 19:35:07 <daemon> Validator balances sum: 96452516167 Gwei
    INFO 2020-12-09 19:35:07 <daemon> Total balance on Beacon: 96452516167000000000 wei
    INFO 2020-12-09 19:35:07 <daemon> Validators on Beacon: 3
    INFO 2020-12-09 19:35:07 <daemon> Time since previous report: 0:07:22 or 442 s
    INFO 2020-12-09 19:35:07 <daemon> Balance change since previus report: +158719204000000000 wei or +0.158719204 ETH
    INFO 2020-12-09 19:35:07 <daemon> Validators change since previus report: +0
    INFO 2020-12-09 19:35:07 <daemon> Tx call data: oracle.reportBeacon(200, 96452516167000000000, 3)
    INFO 2020-12-09 19:35:07 <daemon> Calling tx locally is succeeded.
Tx data: {'value': 0, 'gasPrice': 1000000000, 'chainId': 1337, 'from': '0xF2De2b2dD8881DE68390Da81EDc6Ab93c8e13291', 'gas': 1500000, 'to': '0xcD3db5ca818a645359e09543Cc0e5b7bB9593229', 'data': '0x62eeb73200000000000000000000000000000000000000000000000000000000000000c80000000000000000000000000000000000000000000000053a8c2c6d32a806000000000000000000000000000000000000000000000000000000000000000003'}
Should we sent this TX? [y/n]: n
    INFO 2020-12-09 19:35:16 <daemon> We are in single-iteration mode, so exiting. Set DAEMON=1 env to run in the loop.
```

ongrid commented:
With cosmetic fixes
```
    INFO 2020-12-09 19:58:00 <daemon> Connected to E2E network (PoA engine)
    INFO 2020-12-09 19:58:00 <daemon> Injecting PoA compatibility middleware
    INFO 2020-12-09 19:58:00 <daemon> MEMBER_PRIV_KEY provided, running in transactable (PRODUCTION) mode
    INFO 2020-12-09 19:58:00 <daemon> Member account: 0xF2De2b2dD8881DE68390Da81EDc6Ab93c8e13291
    INFO 2020-12-09 19:58:00 <daemon> DAEMON=0 Running in single iteration mode (will exit after reporting).
    INFO 2020-12-09 19:58:00 <daemon> ETH1_NODE=http://127.0.0.1:8545
    INFO 2020-12-09 19:58:00 <daemon> BEACON_NODE=http://127.0.0.1:5052 (Lighthouse API)
    INFO 2020-12-09 19:58:00 <daemon> SLEEP=60 s (pause between iterations in DAEMON mode)
    INFO 2020-12-09 19:58:00 <daemon> GAS_LIMIT=1500000 gas units
    INFO 2020-12-09 19:58:00 <daemon> POOL_CONTRACT=0xc12e8e7adcEaF31c1Ca5F8aFD99AB88439628183
    INFO 2020-12-09 19:58:00 <daemon> Oracle contract address: 0xcD3db5ca818a645359e09543Cc0e5b7bB9593229 (auto-discovered)
    INFO 2020-12-09 19:58:00 <daemon> Registry contract address: 0x7fAF80E96530e5cd13a1f35701fcc6b334B2FD75 (auto-discovered)
    INFO 2020-12-09 19:58:00 <daemon> Seconds per slot: 1 (auto-discovered)
    INFO 2020-12-09 19:58:00 <daemon> Slots per epoch: 8 (auto-discovered)
    INFO 2020-12-09 19:58:00 <daemon> Epochs per frame: 20 (auto-discovered)
    INFO 2020-12-09 19:58:00 <daemon> Genesis time: 1607529992 (auto-discovered)
    INFO 2020-12-09 19:58:00 <daemon> Starting the main loop
    INFO 2020-12-09 19:58:00 <daemon> Getting previously reported numbers (will be fetched from events)...
    INFO 2020-12-09 19:58:00 <daemon> Previously reported epoch: 140
    INFO 2020-12-09 19:58:00 <daemon> Previously reported beaconBalance: 96293796963000000000 wei or 96.29379696299999 ETH
    INFO 2020-12-09 19:58:00 <daemon> Previously reported beaconValidators: 3
    INFO 2020-12-09 19:58:00 <daemon> Timestamp of previous report: 2020-12-09 19:27:45 or 1607531265
    INFO 2020-12-09 19:58:00 <daemon> Potentially reportable epoch: 380 (from ETH1 contract)
    INFO 2020-12-09 19:58:00 <daemon> Last finalized epoch: 383 (from Beacon)
    INFO 2020-12-09 19:58:00 <daemon> Reportable state: epoch:380 slot:3040
    INFO 2020-12-09 19:58:00 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-09 19:58:00 <daemon> Node operators in registry: 2
    INFO 2020-12-09 19:58:00 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-09 19:58:00 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-09 19:58:00 <daemon> Total validator keys in registry: 20
    INFO 2020-12-09 19:58:00 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-09 19:58:01 <daemon> Validator balances on beacon for slot: 3040
    INFO 2020-12-09 19:58:01 <daemon> Pubkey: 0x94f9f5c1ad Balance: 32317631296 Gwei
    INFO 2020-12-09 19:58:01 <daemon> Pubkey: 0xa3d9511870 Balance: 32312357013 Gwei
    INFO 2020-12-09 19:58:01 <daemon> Pubkey: 0x8e7ebb0d21 Balance: 32313803028 Gwei
    INFO 2020-12-09 19:58:01 <daemon> Lido validators' sum. balance on Beacon: 96943791337000000000 wei or 96.943791337 ETH
    INFO 2020-12-09 19:58:01 <daemon> Lido validators visible on Beacon: 3
    INFO 2020-12-09 19:58:01 <daemon> Time since previous report: 0:30:16 or 1816 s
    INFO 2020-12-09 19:58:01 <daemon> Balance change since previous report: +649994374000000000 wei or +0.649994374 ETH
    INFO 2020-12-09 19:58:01 <daemon> Validators change since previus report: +0
    INFO 2020-12-09 19:58:01 <daemon> Tx call data: oracle.reportBeacon(380, 96943791337000000000, 3)
    INFO 2020-12-09 19:58:01 <daemon> Calling tx locally succeeded.
Tx data: {'value': 0, 'gasPrice': 1000000000, 'chainId': 1337, 'from': '0xF2De2b2dD8881DE68390Da81EDc6Ab93c8e13291', 'gas': 1500000, 'to': '0xcD3db5ca818a645359e09543Cc0e5b7bB9593229', 'data': '0x62eeb732000000000000000000000000000000000000000000000000000000000000017c000000000000000000000000000000000000000000000005415d88973d79da000000000000000000000000000000000000000000000000000000000000000003'}
Should we send this TX? [y/n]: y
    INFO 2020-12-09 19:58:13 <daemon> Preparing TX... CTRL-C to abort
    INFO 2020-12-09 19:58:16 <daemon> TX hash: 0xae24df726349e794d9d19e94beb01d144955372fcbca60d588428158b23ed67a ... CTRL-C to abort
    INFO 2020-12-09 19:58:19 <daemon> Sending TX... CTRL-C to abort
    INFO 2020-12-09 19:58:22 <daemon> TX has been sent. Waiting for receipt...
    INFO 2020-12-09 19:58:26 <daemon> TX successful
    INFO 2020-12-09 19:58:26 <daemon> We are in single-iteration mode, so exiting. Set DAEMON=1 env to run in the loop.
```

ongrid commented:
In daemon mode + started from scratch (there were no prior events on the contract)

```
    INFO 2020-12-09 17:20:17 <daemon> Connected to E2E network (PoA engine)
    INFO 2020-12-09 17:20:17 <daemon> Injecting PoA compatibility middleware
    INFO 2020-12-09 17:20:17 <daemon> MEMBER_PRIV_KEY provided, running in transactable (PRODUCTION) mode
    INFO 2020-12-09 17:20:17 <daemon> Member account: 0x656e544DeaB532e9f5B8b8079B3809Aa1757fb0D
    INFO 2020-12-09 17:20:17 <daemon> DAEMON=1 Running in daemon mode (in endless loop).
    INFO 2020-12-09 17:20:17 <daemon> ETH1_NODE=http://node1:8545
    INFO 2020-12-09 17:20:17 <daemon> BEACON_NODE=http://node2-1:5052 (Lighthouse API)
    INFO 2020-12-09 17:20:17 <daemon> SLEEP=60 s (pause between iterations in DAEMON mode)
    INFO 2020-12-09 17:20:17 <daemon> GAS_LIMIT=1500000 gas units
    INFO 2020-12-09 17:20:17 <daemon> POOL_CONTRACT=0xc12e8e7adcEaF31c1Ca5F8aFD99AB88439628183
    INFO 2020-12-09 17:20:17 <daemon> Oracle contract address: 0xcD3db5ca818a645359e09543Cc0e5b7bB9593229 (auto-discovered)
    INFO 2020-12-09 17:20:17 <daemon> Registry contract address: 0x7fAF80E96530e5cd13a1f35701fcc6b334B2FD75 (auto-discovered)
    INFO 2020-12-09 17:20:17 <daemon> Seconds per slot: 1 (auto-discovered)
    INFO 2020-12-09 17:20:17 <daemon> Slots per epoch: 8 (auto-discovered)
    INFO 2020-12-09 17:20:17 <daemon> Epochs per frame: 20 (auto-discovered)
    INFO 2020-12-09 17:20:17 <daemon> Genesis time: 1607534165 (auto-discovered)
    INFO 2020-12-09 17:20:17 <daemon> Starting the main loop
    INFO 2020-12-09 17:20:17 <daemon> Getting previously reported numbers (will be fetched from events)...
 WARNING 2020-12-09 17:20:17 <daemon> No events on the contract
    INFO 2020-12-09 17:20:17 <daemon> Potentially reportable epoch: 20 (from ETH1 contract)
    INFO 2020-12-09 17:20:17 <daemon> Last finalized epoch: 29 (from Beacon)
    INFO 2020-12-09 17:20:17 <daemon> Reportable state: epoch:20 slot:160
    INFO 2020-12-09 17:20:17 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-09 17:20:17 <daemon> Node operators in registry: 2
    INFO 2020-12-09 17:20:17 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-09 17:20:18 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-09 17:20:18 <daemon> Total validator keys in registry: 20
    INFO 2020-12-09 17:20:18 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-09 17:20:18 <daemon> Validator balances on beacon for slot: 160
    INFO 2020-12-09 17:20:18 <daemon> Lido validators' sum. balance on Beacon: 0 wei or 0.0 ETH
    INFO 2020-12-09 17:20:18 <daemon> Lido validators visible on Beacon: 0
    INFO 2020-12-09 17:20:18 <daemon> Tx call data: oracle.reportBeacon(20, 0, 0)
    INFO 2020-12-09 17:20:18 <daemon> Calling tx locally succeeded.
    INFO 2020-12-09 17:20:18 <daemon> Preparing TX... CTRL-C to abort
    INFO 2020-12-09 17:20:21 <daemon> TX hash: 0x71ced09de4ac57cd01fc28935e9c9bf5bde81a27e9becf52dcf60317649f724a ... CTRL-C to abort
    INFO 2020-12-09 17:20:24 <daemon> Sending TX... CTRL-C to abort
    INFO 2020-12-09 17:20:27 <daemon> TX has been sent. Waiting for receipt...
    INFO 2020-12-09 17:20:32 <daemon> TX successful
    INFO 2020-12-09 17:20:32 <daemon> We are in DAEMON mode. Sleep 60 s and continue
```

ongrid commented:
If tx call has been reverted when played locally (already reported): 

```
    INFO 2020-12-09 17:21:32 <daemon> Getting previously reported numbers (will be fetched from events)...
    INFO 2020-12-09 17:21:32 <daemon> Previously reported epoch: 20
    INFO 2020-12-09 17:21:32 <daemon> Previously reported beaconBalance: 0 wei or 0.0 ETH
    INFO 2020-12-09 17:21:32 <daemon> Previously reported beaconValidators: 0
    INFO 2020-12-09 17:21:32 <daemon> Timestamp of previous report: 2020-12-09 17:20:32 or 1607534432
    INFO 2020-12-09 17:21:32 <daemon> Potentially reportable epoch: 40 (from ETH1 contract)
    INFO 2020-12-09 17:21:32 <daemon> Last finalized epoch: 38 (from Beacon)
    INFO 2020-12-09 17:21:32 <daemon> Reportable state: epoch:20 slot:160
    INFO 2020-12-09 17:21:32 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-09 17:21:32 <daemon> Node operators in registry: 2
    INFO 2020-12-09 17:21:32 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-09 17:21:32 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-09 17:21:32 <daemon> Total validator keys in registry: 20
    INFO 2020-12-09 17:21:32 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-09 17:21:33 <daemon> Validator balances on beacon for slot: 160
    INFO 2020-12-09 17:21:33 <daemon> Lido validators' sum. balance on Beacon: 0 wei or 0.0 ETH
    INFO 2020-12-09 17:21:33 <daemon> Lido validators visible on Beacon: 0
    INFO 2020-12-09 17:21:33 <daemon> Time since previous report: 0:01:01 or 61 s
    INFO 2020-12-09 17:21:33 <daemon> Balance change since previous report: +0 wei or +0.0 ETH
    INFO 2020-12-09 17:21:33 <daemon> Validators change since previus report: +0
    INFO 2020-12-09 17:21:33 <daemon> Tx call data: oracle.reportBeacon(20, 0, 0)
 WARNING 2020-12-09 17:21:33 <daemon> execution reverted: EPOCH_IS_TOO_OLD
    INFO 2020-12-09 17:21:33 <daemon> We are in DAEMON mode. Sleep 60 s and continue
```

The failure reason will be decorated by more friendly and explaining message. Severity will be decreased to INFO. #64

Klyaus commented on app/oracle.py:
```diff
@@ -165,53 +181,82 @@ def prompt(prompt_message, prompt_end):
+    else:
+        logging.warning('No events on the contract')
+        return False
+
```
according to PEP 8: 2 blank lines are expected after class or function definition

Klyaus commented on app/oracle.py:
```diff
@@ -165,53 +181,82 @@ def prompt(prompt_message, prompt_end):
+    logging.info('Getting previously reported numbers (will be fetched from events)...')
+    SECONDS_PER_ETH1_BLOCK = 14
+    latest_block = w3.eth.getBlock('latest')
+    #Calculate earliest block to limit scanning depth
```
according to PEP 8: after '#' need a space

Klyaus commented on app/oracle.py:
```diff
@@ -8,6 +8,8 @@
 import time
 import traceback
+import datetime
+import time
```
duplicate import
and it is worth removing more unused imports

vshvsh commented on app/oracle.py:
```diff
@@ -165,53 +181,82 @@ def prompt(prompt_message, prompt_end):
+    logging.info(f'Lido validators visible on Beacon: {validators}')
+
+    if prev_frame:
+        balance_change = balance - prev_balance
```
use the same logic as in the smart contract: balance change is not the important metric, reward/slashing size is and it's calculated in a way more complex than this
reward should be shown in absolute and relative metrics, along with an APR

vshvsh commented on app/oracle.py:
```diff
@@ -165,53 +181,82 @@ def prompt(prompt_message, prompt_end):
+        logging.info(f'Time since previous report: {datetime.timedelta(seconds = delta_seconds)} or {delta_seconds} s')
+        logging.info(f'Balance change since previous report: {balance_change:+} wei or {(balance_change / 1e18):+} ETH')
+        if balance_change < 0:
+            logging.warning('Balance decreased! Validators were either slashed or suffered penalties!')
```
balance can increase even when slashings or penalites are severe enough to warrant loss for stakers (i.e. when the number of validators increases between reports)

ongrid commented:
@vshvsh 

First report (no prior events). 
```
    INFO 2020-12-10 16:11:23 <daemon> Connected to E2E network (PoA engine)
    INFO 2020-12-10 16:11:23 <daemon> Injecting PoA compatibility middleware
    INFO 2020-12-10 16:11:23 <daemon> MEMBER_PRIV_KEY provided, running in transactable (PRODUCTION) mode
    INFO 2020-12-10 16:11:23 <daemon> Member account: 0x656e544DeaB532e9f5B8b8079B3809Aa1757fb0D
    INFO 2020-12-10 16:11:24 <daemon> DAEMON=1 Running in daemon mode (in endless loop).
    INFO 2020-12-10 16:11:24 <daemon> ETH1_NODE=http://node1:8545
    INFO 2020-12-10 16:11:24 <daemon> BEACON_NODE=http://node2-1:5052 (Lighthouse API)
    INFO 2020-12-10 16:11:24 <daemon> SLEEP=60 s (pause between iterations in DAEMON mode)
    INFO 2020-12-10 16:11:24 <daemon> GAS_LIMIT=1500000 gas units
    INFO 2020-12-10 16:11:24 <daemon> POOL_CONTRACT=0xc12e8e7adcEaF31c1Ca5F8aFD99AB88439628183
    INFO 2020-12-10 16:11:24 <daemon> Oracle contract address: 0xcD3db5ca818a645359e09543Cc0e5b7bB9593229 (auto-discovered)
    INFO 2020-12-10 16:11:24 <daemon> Registry contract address: 0x7fAF80E96530e5cd13a1f35701fcc6b334B2FD75 (auto-discovered)
    INFO 2020-12-10 16:11:24 <daemon> Seconds per slot: 1 (auto-discovered)
    INFO 2020-12-10 16:11:24 <daemon> Slots per epoch: 8 (auto-discovered)
    INFO 2020-12-10 16:11:24 <daemon> Epochs per frame: 20 (auto-discovered)
    INFO 2020-12-10 16:11:24 <daemon> Genesis time: 1607616355 (auto-discovered)
    INFO 2020-12-10 16:11:24 <daemon> Starting the main loop
    INFO 2020-12-10 16:11:24 <daemon> Getting previously reported numbers (will be fetched from events)...
    INFO 2020-12-10 16:11:24 <daemon> No events on the contract. It's ok if it's the first run.
    INFO 2020-12-10 16:11:24 <daemon> Potentially reportable epoch: 40 (from ETH1 contract)
    INFO 2020-12-10 16:11:24 <daemon> Last finalized epoch: 39 (from Beacon)
    INFO 2020-12-10 16:11:24 <daemon> Reportable state: epoch:20 slot:160
    INFO 2020-12-10 16:11:24 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-10 16:11:24 <daemon> Node operators in registry: 2
    INFO 2020-12-10 16:11:24 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-10 16:11:24 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-10 16:11:24 <daemon> Total validator keys in registry: 20
    INFO 2020-12-10 16:11:24 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-10 16:11:24 <daemon> Validator balances on beacon for slot: 160
    INFO 2020-12-10 16:11:24 <daemon> Lido validators' sum. balance on Beacon: 0 wei or 0.0 ETH
    INFO 2020-12-10 16:11:24 <daemon> Lido validators visible on Beacon: 0
    INFO 2020-12-10 16:11:24 <daemon> Tx call data: oracle.reportBeacon(20, 0, 0)
    INFO 2020-12-10 16:11:25 <daemon> Calling tx locally succeeded.
    INFO 2020-12-10 16:11:25 <daemon> Preparing TX... CTRL-C to abort
    INFO 2020-12-10 16:11:28 <daemon> TX hash: 0x71ced09de4ac57cd01fc28935e9c9bf5bde81a27e9becf52dcf60317649f724a ... CTRL-C to abort
    INFO 2020-12-10 16:11:31 <daemon> Sending TX... CTRL-C to abort
    INFO 2020-12-10 16:11:34 <daemon> TX has been sent. Waiting for receipt...
    INFO 2020-12-10 16:11:39 <daemon> TX successful
    INFO 2020-12-10 16:11:39 <daemon> We are in DAEMON mode. Sleep 60 s and continue
```

Second report, without rewards
```
    INFO 2020-12-10 16:12:38 <daemon> Getting previously reported numbers (will be fetched from events)...
    INFO 2020-12-10 16:12:39 <daemon> Previously reported epoch: 20
    INFO 2020-12-10 16:12:39 <daemon> Previously reported beaconBalance: 0 wei or 0.0 ETH
    INFO 2020-12-10 16:12:39 <daemon> Previously reported bufferedBalance: 27000000000000000000 wei or 27.0 ETH
    INFO 2020-12-10 16:12:39 <daemon> Previous validator metrics: depositedValidators:3
    INFO 2020-12-10 16:12:39 <daemon> Previous validator metrics: transientValidators:3
    INFO 2020-12-10 16:12:39 <daemon> Previous validator metrics: beaconValidators:0
    INFO 2020-12-10 16:12:39 <daemon> Timestamp of previous report: 2020-12-10 16:11:39 or 1607616699
    INFO 2020-12-10 16:12:39 <daemon> Potentially reportable epoch: 40 (from ETH1 contract)
    INFO 2020-12-10 16:12:39 <daemon> Last finalized epoch: 48 (from Beacon)
    INFO 2020-12-10 16:12:39 <daemon> Reportable state: epoch:40 slot:320
    INFO 2020-12-10 16:12:39 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-10 16:12:39 <daemon> Node operators in registry: 2
    INFO 2020-12-10 16:12:39 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-10 16:12:39 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-10 16:12:39 <daemon> Total validator keys in registry: 20
    INFO 2020-12-10 16:12:39 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-10 16:12:39 <daemon> Validator balances on beacon for slot: 320
    INFO 2020-12-10 16:12:39 <daemon> Lido validators' sum. balance on Beacon: 0 wei or 0.0 ETH
    INFO 2020-12-10 16:12:39 <daemon> Lido validators visible on Beacon: 0
    INFO 2020-12-10 16:12:39 <daemon> Time delta: 0:01:00 or 60 s
    INFO 2020-12-10 16:12:39 <daemon> depositedValidators before:3 after:3 change:0
    INFO 2020-12-10 16:12:39 <daemon> beaconValidators before:0 after:0 change:0
    INFO 2020-12-10 16:12:39 <daemon> transientValidators before:3 after:3 change:0
    INFO 2020-12-10 16:12:39 <daemon> beaconBalance before:0 after:0 change:0
    INFO 2020-12-10 16:12:39 <daemon> bufferedBalance before:27000000000000000000 after:27000000000000000000 change:0
    INFO 2020-12-10 16:12:39 <daemon> transientBalance before:96000000000000000000 after:96000000000000000000 change:0
    INFO 2020-12-10 16:12:39 <daemon> totalPooledEther before:123000000000000000000 after:123000000000000000000 
    INFO 2020-12-10 16:12:39 <daemon> totalPooledEther will stay intact. This won't have any economical impact on the pool.
    INFO 2020-12-10 16:12:39 <daemon> Tx call data: oracle.reportBeacon(40, 0, 0)
    INFO 2020-12-10 16:12:39 <daemon> Calling tx locally succeeded.
    INFO 2020-12-10 16:12:39 <daemon> Preparing TX... CTRL-C to abort
    INFO 2020-12-10 16:12:42 <daemon> TX hash: 0x401537c9b76646f924eeb4da80ab5dae256c4b0e48bb34138db0898114ee5721 ... CTRL-C to abort
    INFO 2020-12-10 16:12:45 <daemon> Sending TX... CTRL-C to abort
    INFO 2020-12-10 16:12:48 <daemon> TX has been sent. Waiting for receipt...
    INFO 2020-12-10 16:12:54 <daemon> TX successful
    INFO 2020-12-10 16:12:54 <daemon> We are in DAEMON mode. Sleep 60 s and continue
```

If epoch has been already reported
```
    INFO 2020-12-10 16:12:54 <daemon> We are in DAEMON mode. Sleep 60 s and continue
    INFO 2020-12-10 16:13:54 <daemon> Getting previously reported numbers (will be fetched from events)...
    INFO 2020-12-10 16:13:54 <daemon> Previously reported epoch: 40
    INFO 2020-12-10 16:13:54 <daemon> Previously reported beaconBalance: 0 wei or 0.0 ETH
    INFO 2020-12-10 16:13:54 <daemon> Previously reported bufferedBalance: 27000000000000000000 wei or 27.0 ETH
    INFO 2020-12-10 16:13:54 <daemon> Previous validator metrics: depositedValidators:3
    INFO 2020-12-10 16:13:54 <daemon> Previous validator metrics: transientValidators:3
    INFO 2020-12-10 16:13:54 <daemon> Previous validator metrics: beaconValidators:0
    INFO 2020-12-10 16:13:54 <daemon> Timestamp of previous report: 2020-12-10 16:12:54 or 1607616774
    INFO 2020-12-10 16:13:54 <daemon> Potentially reportable epoch: 40 (from ETH1 contract)
    INFO 2020-12-10 16:13:54 <daemon> Last finalized epoch: 57 (from Beacon)
    INFO 2020-12-10 16:13:54 <daemon> Reportable state: epoch:40 slot:320
    INFO 2020-12-10 16:13:54 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-10 16:13:54 <daemon> Node operators in registry: 2
    INFO 2020-12-10 16:13:54 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-10 16:13:54 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-10 16:13:54 <daemon> Total validator keys in registry: 20
    INFO 2020-12-10 16:13:54 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-10 16:13:54 <daemon> Validator balances on beacon for slot: 320
    INFO 2020-12-10 16:13:54 <daemon> Lido validators' sum. balance on Beacon: 0 wei or 0.0 ETH
    INFO 2020-12-10 16:13:54 <daemon> Lido validators visible on Beacon: 0
    INFO 2020-12-10 16:13:54 <daemon> Currently reportable epoch 40 has already been reported. Skipping it.
    INFO 2020-12-10 16:13:54 <daemon> We are in DAEMON mode. Sleep 60 s and continue
```

When reward expected 
```
    INFO 2020-12-10 16:13:54 <daemon> Getting previously reported numbers (will be fetched from events)...
    INFO 2020-12-10 16:13:54 <daemon> Previously reported epoch: 40
    INFO 2020-12-10 16:13:54 <daemon> Previously reported beaconBalance: 0 wei or 0.0 ETH
    INFO 2020-12-10 16:13:54 <daemon> Previously reported bufferedBalance: 27000000000000000000 wei or 27.0 ETH
    INFO 2020-12-10 16:13:54 <daemon> Previous validator metrics: depositedValidators:3
    INFO 2020-12-10 16:13:54 <daemon> Previous validator metrics: transientValidators:3
    INFO 2020-12-10 16:13:54 <daemon> Previous validator metrics: beaconValidators:0
    INFO 2020-12-10 16:13:54 <daemon> Timestamp of previous report: 2020-12-10 16:12:54 or 1607616774
    INFO 2020-12-10 16:13:54 <daemon> Potentially reportable epoch: 40 (from ETH1 contract)
    INFO 2020-12-10 16:13:54 <daemon> Last finalized epoch: 57 (from Beacon)
    INFO 2020-12-10 16:13:54 <daemon> Reportable state: epoch:40 slot:320
    INFO 2020-12-10 16:13:54 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-10 16:13:54 <daemon> Node operators in registry: 2
    INFO 2020-12-10 16:13:54 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-10 16:13:54 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-10 16:13:54 <daemon> Total validator keys in registry: 20
    INFO 2020-12-10 16:13:54 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-10 16:13:54 <daemon> Validator balances on beacon for slot: 320
    INFO 2020-12-10 16:13:54 <daemon> Lido validators' sum. balance on Beacon: 0 wei or 0.0 ETH
    INFO 2020-12-10 16:13:54 <daemon> Lido validators visible on Beacon: 0
    INFO 2020-12-10 16:13:54 <daemon> Currently reportable epoch 40 has already been reported. Skipping it.
    INFO 2020-12-10 16:13:54 <daemon> We are in DAEMON mode. Sleep 60 s and continue
    INFO 2020-12-10 16:14:54 <daemon> Getting previously reported numbers (will be fetched from events)...
    INFO 2020-12-10 16:14:54 <daemon> Previously reported epoch: 40
    INFO 2020-12-10 16:14:54 <daemon> Previously reported beaconBalance: 0 wei or 0.0 ETH
    INFO 2020-12-10 16:14:54 <daemon> Previously reported bufferedBalance: 27000000000000000000 wei or 27.0 ETH
    INFO 2020-12-10 16:14:54 <daemon> Previous validator metrics: depositedValidators:3
    INFO 2020-12-10 16:14:54 <daemon> Previous validator metrics: transientValidators:3
    INFO 2020-12-10 16:14:54 <daemon> Previous validator metrics: beaconValidators:0
    INFO 2020-12-10 16:14:54 <daemon> Timestamp of previous report: 2020-12-10 16:12:54 or 1607616774
    INFO 2020-12-10 16:14:54 <daemon> Potentially reportable epoch: 60 (from ETH1 contract)
    INFO 2020-12-10 16:14:54 <daemon> Last finalized epoch: 65 (from Beacon)
    INFO 2020-12-10 16:14:54 <daemon> Reportable state: epoch:60 slot:480
    INFO 2020-12-10 16:14:54 <daemon> Quering NodeOperatorsRegistry...
    INFO 2020-12-10 16:14:54 <daemon> Node operators in registry: 2
    INFO 2020-12-10 16:14:54 <daemon> Node operator ID: 0 Keys: 10
    INFO 2020-12-10 16:14:54 <daemon> Node operator ID: 1 Keys: 10
    INFO 2020-12-10 16:14:55 <daemon> Total validator keys in registry: 20
    INFO 2020-12-10 16:14:55 <daemon> Fetching validators from Beacon node...
    INFO 2020-12-10 16:14:56 <daemon> Validator balances on beacon for slot: 480
    INFO 2020-12-10 16:14:56 <daemon> Pubkey: 0x8e7ebb0d21 Balance: 32006485704 Gwei
    INFO 2020-12-10 16:14:56 <daemon> Pubkey: 0x94f9f5c1ad Balance: 32005363114 Gwei
    INFO 2020-12-10 16:14:56 <daemon> Pubkey: 0xa3d9511870 Balance: 32006485704 Gwei
    INFO 2020-12-10 16:14:56 <daemon> Lido validators' sum. balance on Beacon: 96018334522000000000 wei or 96.018334522 ETH
    INFO 2020-12-10 16:14:56 <daemon> Lido validators visible on Beacon: 3
    INFO 2020-12-10 16:14:56 <daemon> Time delta: 0:02:00 or 120 s
    INFO 2020-12-10 16:14:56 <daemon> depositedValidators before:3 after:3 change:0
    INFO 2020-12-10 16:14:56 <daemon> beaconValidators before:0 after:3 change:3
    INFO 2020-12-10 16:14:56 <daemon> transientValidators before:3 after:0 change:-3
    INFO 2020-12-10 16:14:56 <daemon> beaconBalance before:0 after:96018334522000000000 change:96018334522000000000
    INFO 2020-12-10 16:14:56 <daemon> bufferedBalance before:27000000000000000000 after:27000000000000000000 change:0
    INFO 2020-12-10 16:14:56 <daemon> transientBalance before:96000000000000000000 after:0 change:-96000000000000000000
    INFO 2020-12-10 16:14:56 <daemon> totalPooledEther before:123000000000000000000 after:123018334522000000000 
    INFO 2020-12-10 16:14:56 <daemon> totalPooledEther will increase by 18334522000000000 wei or 0.018334522 ETH
    INFO 2020-12-10 16:14:56 <daemon> Increase since last report: 0.0149 %
    INFO 2020-12-10 16:14:56 <daemon> Expected APR: 3917.3271 %
    INFO 2020-12-10 16:14:56 <daemon> Tx call data: oracle.reportBeacon(60, 96018334522000000000, 3)
    INFO 2020-12-10 16:14:56 <daemon> Calling tx locally succeeded.
    INFO 2020-12-10 16:14:56 <daemon> Preparing TX... CTRL-C to abort
    INFO 2020-12-10 16:14:59 <daemon> TX hash: 0x4bed1260125d868a513efb00f5b1b19dc2279baedd619812079103c1d51234bf ... CTRL-C to abort
    INFO 2020-12-10 16:15:02 <daemon> Sending TX... CTRL-C to abort
    INFO 2020-12-10 16:15:05 <daemon> TX has been sent. Waiting for receipt...
    INFO 2020-12-10 16:15:09 <daemon> TX successful
    INFO 2020-12-10 16:15:09 <daemon> We are in DAEMON mode. Sleep 60 s and continue
```
